packages/core-data/src/types/user.ts

```diff
@@ -0,0 +1,83 @@
+/**
+ * Internal dependencies
+ */
+import { AvatarUrls, Context, ContextualField } from './common';
+
+export interface User< C extends Context > {
+	/**
+	 * Unique identifier for the user.
+	 */
+	id: number;
+	/**
+	 * Login name for the user.
+	 */
+	username: ContextualField< string, 'edit', C >;
+	/**
+	 * Display name for the user.
+	 */
+	name: string;
+	/**
+	 * First name for the user.
+	 */
+	first_name: ContextualField< string, 'edit', C >;
+	/**
+	 * Last name for the user.
+	 */
+	last_name: ContextualField< string, 'edit', C >;
+	/**
+	 * The email address for the user.
+	 */
+	email: ContextualField< string, 'edit', C >;
+	/**
+	 * URL of the user.
+	 */
+	url: string;
+	/**
+	 * Description of the user.
+	 */
+	description: string;
+	/**
+	 * Author URL of the user.
+	 */
+	link: string;
+	/**
+	 * Locale for the user.
+	 */
+	locale: ContextualField< '' | 'en_US', 'edit', C >;
+	/**
+	 * The nickname for the user.
+	 */
+	nickname: ContextualField< string, 'edit', C >;
+	/**
+	 * An alphanumeric identifier for the user.
+	 */
+	slug: string;
+	/**
+	 * Registration date for the user.
+	 */
+	registered_date: ContextualField< string, 'edit', C >;
+	/**
+	 * Roles assigned to the user.
+	 */
+	roles: ContextualField< string[], 'edit', C >;
+	/**
+	 * Password for the user (never included).
+	 */
+	password: string;
```

Comment:
if this is never included then it shouldn't be asserted that it exists. I am assuming we have it here because the `Updatable<User>` needs it?

if so, there are two ways we could handle this:

We can set it as optional and expand in the comment that it's never received but exists because you can set it when updating.

```ts
/**
 * Password for the user.
 *
 * This is never sent from the server to the client
 * but exists because we might send an update to the
 * server with a new password to set.
 */
password?: string;
```

Or we could introduce a new context value, `update`, that will include this for `Updatable` but leave it for the other cases.

```ts
export type ContextualField<
 	FieldType,
 	AvailableInContexts extends Context,
 	C extends Context
 > = C extends 'update' ? (
     AvailableInContexts extends C ? FieldType : (
         AvailableInContexts extends 'edit' ? FieldType : never
    )
 ) : C extends AvailableInContexts ? FieldType : never;
```

Look at [the playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAwg9gO2BAHsKBeKByCATAS2GygB8cBXMPAQ2RPOwDcCIB3BnCAWwCN9sAbgBQw1GDgAndKEixEyNBRoAbAGKsVeADzCoASA0QtAFXAQANHv0BBJjQIqavFRACSCeElTAAzlB8IBDx-L0VgKwMYALQgkPlvND0APkxYGORg-2wqWnooAH4oAAo9KHKoOwcnF3dPBR9-QKz0oqNTcygALhKyiv6qx2dXDzDGjLjs-CISNs08MzkehAgmCEk+qABKPS3u9Ob4wZqR+sS-Qqh2hc7l1fWRYQBjRF90GjSAbwBfKBpQhpKVTXbRvSQEBAAcwsXEIxBhuDh2GSIhowgA9OioAA9AqiF4IN5QXhfX7-BLhZTqeag4DgqEIljsbAI6bEFHCXgYrG4-GvdBPUl-AHnKkgsEQ6GUah0CAs2EzDlPbk4vHPflQPBC8ljIHU4w6CUMhXw6V5OUcvAq3nqwnoaBYH7Cik+MU0o1S5isDgI3Ky5EiCDWtVAA) for examples.

Reply:
Let's go with an optional field for now, I'd rather not introduce entirely new mechanics for a single field that isn't even used at the moment. The new context seems confusing to me, the API only has three of them and the fourth would be a special client-only construct. I wonder if we could have a special type like `UpdateOnlyField` that would be `never` by default, but would get "activated" by the `Updatable` type wrapper.

Reply:
👍 to all of that.

sometimes we have an awkward type, which is okay and something we can improve later. this is 100x better than having an incorrect type.